src/imitation/algorithms/sqil.py

```diff
@@ -0,0 +1,215 @@
+        """
+        super().__init__(
+            buffer_size,
+            observation_space,
+            action_space,
+            device,
+            n_envs,
+            optimize_memory_usage,
+            handle_timeout_termination=False,
+        )
+
+        self.set_demonstrations(demonstrations)
+
+    def set_demonstrations(
+        self,
+        demonstrations: algo_base.AnyTransitions,
+    ) -> None:
+        """Set the demonstrations to be used in the buffer.
+
+        Args:
+            demonstrations (algo_base.AnyTransitions): Expert demonstrations.
+        """
+        # If demonstrations is a list of trajectories,
+        # flatten it into a list of transitions
+        if not isinstance(demonstrations, types.TransitionsMinimal):
+            (
+                item,
+                demonstrations,
+            ) = util.get_first_iter_element(  # type: ignore[assignment]
+                demonstrations,  # type: ignore[assignment]
+            )
+            if isinstance(item, types.Trajectory):
+                demonstrations = rollout.flatten_trajectories(
+                    demonstrations,  # type: ignore[arg-type]
+                )
+            else:  # item is a TransitionMapping
+                demonstrations = rollout.flatten_transition_mappings(
+                    demonstrations,  # type: ignore[arg-type]
+                )
+
+        assert isinstance(demonstrations, types.Transitions)
+
+        n_samples = len(demonstrations)
+        self.expert_buffer = buffers.ReplayBuffer(
+            n_samples,
+            self.observation_space,
+            self.action_space,
+            handle_timeout_termination=False,
+        )
+
+        for transition in demonstrations:
+            self.expert_buffer.add(
+                obs=np.array(transition["obs"]),
+                next_obs=np.array(transition["next_obs"]),
+                action=np.array(transition["acts"]),
+                done=np.array(transition["dones"]),
+                reward=np.array(1),
+                infos=[{}],
+            )
+
+    def sample(
+        self,
+        batch_size: int,
+        env: Optional[vec_env.VecNormalize] = None,
+    ) -> buffers.ReplayBufferSamples:
+        """Sample a batch of data.
+
+        Half of the batch will be from the expert buffer,
+        and the other half will be from the collected data.
+
+        Args:
+            batch_size: Number of element to sample in total
+            env: associated gym VecEnv to normalize the observations/rewards
+                when sampling
+
+        Returns:
+            A batch of samples for DQN
+        """
+        new_sample_size, expert_sample_size = util.split_in_half(batch_size)
+
+        new_sample = super().sample(new_sample_size, env)
+        new_sample.rewards.fill_(0)
+
+        expert_sample = self.expert_buffer.sample(expert_sample_size, env)
+        expert_sample.rewards.fill_(1)
```

Comment:
I think we already do this in `set_demonstrations()` so either remove it there (my preference) or here.